File: weekly-contest/weekly-contest-153/3-maximum-subarray-sum-with-one-deletion/bottom-up-DP-v2.js
/**
 * Bottom-up DP（空间优化版本）
 * 
 * 时间：`O(N)`, 76ms
 * 空间：`O(1)`, 39MB
 */
var maximumSum = function (arr) {
  // 因为至少要选一个，所以取第一个元素，作为初始值
  let [one, zero] = [arr[0], -Infinity]
  let res = Math.max(one, zero)

  for (let i = 1; i < arr.length; ++i) {
    zero = Math.max(
      arr[i], // 单独
      zero + arr[i], // 接上
      one // 接上，但删除本元素
    )
    // 注：这里必须将one放在zero后更新，因为上面的zero需要用到旧的one
    // 或者如果用滚动数组，就不必在乎
    one = Math.max(
      arr[i], // 单独
      one + arr[i] // 接上
    )

    res = Math.max(res, zero, one)
  }

  return res
};

module.exports = maximumSum

File: weekly-contest/weekly-contest-174/4-jump-game-v/BottomupDP-v2.cpp
/**
 * Bottom-up DP：dp[i]表示从`i`跳出去的最大访问数
 * 
 * 时间：`O(NlogN + ND)`, 32ms
 * 空间：`O(N)`
 */

class Solution {
public:
    int maxJumps(vector<int>& arr, int d) {
        const int n = arr.size();

        // 按照高度从低到高排序
        vector<int> ids;
        for (int i = 0; i < n; ++i) ids.push_back(i);
        sort(ids.begin(), ids.end(), [&](int i, int j) {
            return arr[i] < arr[j];
        });

        // 状态表示：dp[i]表示从`i`跳出去的最大访问数
        vector<int> dp(n, 1);
        int res = 1;
        for (int i : ids) {
            // 从`i`跳到左边更低的`j`
            for (int j = i - 1; j >= 0 && i - j <= d && arr[j] < arr[i]; --j) {
                dp[i] = max(dp[i], dp[j] + 1);
            }

            // 从`i`跳到右边更低的`j`
            for (int j = i + 1; j < n && j - i <= d && arr[j] < arr[i]; ++j) {
                dp[i] = max(dp[i], dp[j] + 1);
            }

            res = max(res, dp[i]);
        }
        return res;
    }
};